### ISSUE TITLE
Keccak build failures on macOS Intel with gcc-11

### ISSUE THREAD

**Issue by dstebila** [open-quantum-safe/liboqs#1166]
Building liboqs with gcc-11 on an Apple Intel Mac yields build failures related to the Keccak xkcp_low code.

Reproduce by checking out main, and then running `cmake -GNinja -DCMAKE_C_COMPILER=gcc-11 ..`.

```
[1354/2353] Building ASM object src/common/sha3/xkcp_low/CMakeFiles/xkcp_low_keccakp_1600_avx2.dir/KeccakP-1600/avx2/KeccakP-1600-AVX2.S.
FAILED: src/common/sha3/xkcp_low/CMakeFiles/xkcp_low_keccakp_1600_avx2.dir/KeccakP-1600/avx2/KeccakP-1600-AVX2.S.o 
/usr/local/bin/gcc-11  -I/Users/dstebila/Dev/OQS/liboqs/build/include -fPIC -march=native -Werror -Wall -Wextra -Wpedantic -Wstrict-prototypes -Wshadow -Wformat=2 -Wfloat-equal -Wwrite-strings -O3 -fomit-frame-pointer -fdata-sections -ffunction-sections -Wl,-dead_strip -Wbad-function-cast -Wa,-defsym,macOS=1 -MD -MT src/common/sha3/xkcp_low/CMakeFiles/xkcp_low_keccakp_1600_avx2.dir/KeccakP-1600/avx2/KeccakP-1600-AVX2.S.o -MF src/common/sha3/xkcp_low/CMakeFiles/xkcp_low_keccakp_1600_avx2.dir/KeccakP-1600/avx2/KeccakP-1600-AVX2.S.o.d -o src/common/sha3/xkcp_low/CMakeFiles/xkcp_low_keccakp_1600_avx2.dir/KeccakP-1600/avx2/KeccakP-1600-AVX2.S.o -c /Users/dstebila/Dev/OQS/liboqs/src/common/sha3/xkcp_low/KeccakP-1600/avx2/KeccakP-1600-AVX2.S
clang: error: no such file or directory: 'macOS=1'
[1367/2353] Building C object src/common/sha3/xkcp_low/CMakeFiles/x...s4_avx2.dir/KeccakP-1600times4/avx2/KeccakP-1600-times4-SIMD256.c.o
ninja: build stopped: subcommand failed.
```

**Comment by baentsch** [comment]
> Reproduce by checking out main, and then running cmake -GNinja -DCMAKE_C_COMPILER=gcc-11 ...

Is `/usr/local/bin/` as shown in the trace above indeed the location of the homebrew installs? What's the output of `gcc-11 --version`?

**Comment by dstebila** [comment]
> > Reproduce by checking out main, and then running cmake -GNinja -DCMAKE_C_COMPILER=gcc-11 ...
> 
> Is `/usr/local/bin/` as shown in the trace above indeed the location of the homebrew installs? What's the output of `gcc-11 --version`?

Yes, `/usr/local/bin/gcc-11` is the Homebrew installation of gcc.  The output of `gcc-11 --version` is `gcc-11 (Homebrew GCC 11.2.0_3) 11.2.0`.

**Comment by dstebila** [comment]
But I am perplexed by the output `clang: error: no such file or directory: 'macOS=1'` -- I don't understand how/why it's calling out to clang.

**Comment by baentsch** [comment]
> I don't understand how/why it's calling out to clang.

Me neither. That was the reason for my question above. What happens when you run the failing command manually? (`/usr/local/bin/gcc-11  -I/Users/dstebila/Dev/OQS/liboqs/build/include -fPIC -march=native -Werror -Wall -Wextra -Wpedantic -Wstrict-prototypes -Wshadow -Wformat=2 -Wfloat-equal -Wwrite-strings -O3 -fomit-frame-pointer -fdata-sections -ffunction-sections -Wl,-dead_strip -Wbad-function-cast -Wa,-defsym,macOS=1 -MD -MT src/common/sha3/xkcp_low/CMakeFiles/xkcp_low_keccakp_1600_avx2.dir/KeccakP-1600/avx2/KeccakP-1600-AVX2.S.o -MF src/common/sha3/xkcp_low/CMakeFiles/xkcp_low_keccakp_1600_avx2.dir/KeccakP-1600/avx2/KeccakP-1600-AVX2.S.o.d -o src/common/sha3/xkcp_low/CMakeFiles/xkcp_low_keccakp_1600_avx2.dir/KeccakP-1600/avx2/KeccakP-1600-AVX2.S.o -c /Users/dstebila/Dev/OQS/liboqs/src/common/sha3/xkcp_low/KeccakP-1600/avx2/KeccakP-1600-AVX2.S`)

-> Same reference to `clang`?? What's the output of (the OSX-equivalent of) `ldd` on `/usr/local/bin/gcc-11`?

**Comment by dstebila** [comment]
The output of the above command (with some extra `-v` in there to increase verbosity) is as follows:

```
/usr/local/bin/gcc-11 -I/Users/dstebila/Dev/OQS/liboqs/build/include -fPIC -v -v -v -march=native -Werror -Wall -Wextra -Wpedantic -Wstrict-prototypes -Wshadow -Wformat=2 -Wfloat-equal -Wwrite-strings -O3 -fomit-frame-pointer -fdata-sections -ffunction-sections -Wl,-dead_strip -Wbad-function-cast -Wa,-defsym,macOS=1 -x assembler-with-cpp -MD -MT src/common/sha3/xkcp_low/CMakeFiles/xkcp_low_keccakp_1600_avx2.dir/KeccakP-1600/avx2/KeccakP-1600-AVX2.S.o -MF src/common/sha3/xkcp_low/CMakeFiles/xkcp_low_keccakp_1600_avx2.dir/KeccakP-1600/avx2/KeccakP-1600-AVX2.S.o.d -o src/common/sha3/xkcp_low/CMakeFiles/xkcp_low_keccakp_1600_avx2.dir/KeccakP-1600/avx2/KeccakP-1600-AVX2.S.o -c /Users/dstebila/Dev/OQS/liboqs/src/common/sha3/xkcp_low/KeccakP-1600/avx2/KeccakP-1600-AVX2.S
Using built-in specs.
COLLECT_AS_OPTIONS='-defsym' 'macOS=1'
COLLECT_GCC=/usr/local/bin/gcc-11
Target: x86_64-apple-darwin21
Configured with: ../configure --prefix=/usr/local/Cellar/gcc/11.2.0_3 --libdir=/usr/local/Cellar/gcc/11.2.0_3/lib/gcc/11 --disable-nls --enable-checking=release --with-gcc-major-version-only --enable-languages=c,c++,objc,obj-c++,fortran,d --program-suffix=-11 --with-gmp=/usr/local/opt/gmp --with-mpfr=/usr/local/opt/mpfr --with-mpc=/usr/local/opt/libmpc --with-isl=/usr/local/opt/isl --with-zstd=/usr/local/opt/zstd --with-pkgversion='Homebrew GCC 11.2.0_3' --with-bugurl=https://github.com/Homebrew/homebrew-core/issues --enable-libphobos --build=x86_64-apple-darwin21 --with-system-zlib --disable-multilib --without-build-config --with-native-system-header-dir=/usr/include --with-sysroot=/Library/Developer/CommandLineTools/SDKs/MacOSX12.sdk
Thread model: posix
Supported LTO compression algorithms: zlib zstd
gcc version 11.2.0 (Homebrew GCC 11.2.0_3) 
COLLECT_GCC_OPTIONS='-I' '/Users/dstebila/Dev/OQS/liboqs/build/include' '-fPIC' '-v' '-v' '-v' '-march=native' '-Werror' '-Wall' '-Wextra' '-Wpedantic' '-Wstrict-prototypes' '-Wshadow' '-Wformat=2' '-Wfloat-equal' '-Wwrite-strings' '-O3' '-fomit-frame-pointer' '-fdata-sections' '-ffunction-sections' '-Wbad-function-cast' '-MD' '-MT' 'src/common/sha3/xkcp_low/CMakeFiles/xkcp_low_keccakp_1600_avx2.dir/KeccakP-1600/avx2/KeccakP-1600-AVX2.S.o' '-MF' 'src/common/sha3/xkcp_low/CMakeFiles/xkcp_low_keccakp_1600_avx2.dir/KeccakP-1600/avx2/KeccakP-1600-AVX2.S.o.d' '-o' 'src/common/sha3/xkcp_low/CMakeFiles/xkcp_low_keccakp_1600_avx2.dir/KeccakP-1600/avx2/KeccakP-1600-AVX2.S.o' '-c' '-mmacosx-version-min=12.0.0' '-asm_macosx_version_min=12.0' '-dumpdir' 'src/common/sha3/xkcp_low/CMakeFiles/xkcp_low_keccakp_1600_avx2.dir/KeccakP-1600/avx2/'
 /usr/local/Cellar/gcc/11.2.0_3/libexec/gcc/x86_64-apple-darwin21/11/cc1 -E -lang-asm -quiet -v -v -v -I /Users/dstebila/Dev/OQS/liboqs/build/include -MD src/common/sha3/xkcp_low/CMakeFiles/xkcp_low_keccakp_1600_avx2.dir/KeccakP-1600/avx2/KeccakP-1600-AVX2.S.d -MF src/common/sha3/xkcp_low/CMakeFiles/xkcp_low_keccakp_1600_avx2.dir/KeccakP-1600/avx2/KeccakP-1600-AVX2.S.o.d -MT src/common/sha3/xkcp_low/CMakeFiles/xkcp_low_keccakp_1600_avx2.dir/KeccakP-1600/avx2/KeccakP-1600-AVX2.S.o -D__DYNAMIC__ /Users/dstebila/Dev/OQS/liboqs/src/common/sha3/xkcp_low/KeccakP-1600/avx2/KeccakP-1600-AVX2.S -march=skylake -mmmx -mpopcnt -msse -msse2 -msse3 -mssse3 -msse4.1 -msse4.2 -mavx -mavx2 -mno-sse4a -mno-fma4 -mno-xop -mfma -mno-avx512f -mbmi -mbmi2 -maes -mpclmul -mno-avx512vl -mno-avx512bw -mno-avx512dq -mno-avx512cd -mno-avx512er -mno-avx512pf -mno-avx512vbmi -mno-avx512ifma -mno-avx5124vnniw -mno-avx5124fmaps -mno-avx512vpopcntdq -mno-avx512vbmi2 -mno-gfni -mno-vpclmulqdq -mno-avx512vnni -mno-avx512bitalg -mno-avx512bf16 -mno-avx512vp2intersect -mno-3dnow -madx -mabm -mno-cldemote -mclflushopt -mno-clwb -mno-clzero -mcx16 -mno-enqcmd -mf16c -mfsgsbase -mfxsr -mhle -msahf -mno-lwp -mlzcnt -mmovbe -mno-movdir64b -mno-movdiri -mno-mwaitx -mno-pconfig -mno-pku -mno-prefetchwt1 -mprfchw -mno-ptwrite -mno-rdpid -mrdrnd -mrdseed -mrtm -mno-serialize -msgx -mno-sha -mno-shstk -mno-tbm -mno-tsxldtrk -mno-vaes -mno-waitpkg -mno-wbnoinvd -mxsave -mxsavec -mxsaveopt -mxsaves -mno-amx-tile -mno-amx-int8 -mno-amx-bf16 -mno-uintr -mno-hreset -mno-kl -mno-widekl -mno-avxvnni --param l1-cache-size=32 --param l1-cache-line-size=64 --param l2-cache-size=12288 -mtune=skylake -fPIC -mmacosx-version-min=12.0.0 -Werror -Wall -Wextra -Wpedantic -Wstrict-prototypes -Wshadow -Wformat=2 -Wfloat-equal -Wwrite-strings -Wbad-function-cast -fPIC -fomit-frame-pointer -fdata-sections -ffunction-sections -O3 -fno-directives-only -o /var/folders/xc/jknbkyc944z3m9knl6xljf0m0000gn/T//ccrgHf0V.s
ignoring nonexistent directory "/Library/Developer/CommandLineTools/SDKs/MacOSX12.sdk/usr/local/include"
ignoring nonexistent directory "/usr/local/Cellar/gcc/11.2.0_3/lib/gcc/11/gcc/x86_64-apple-darwin21/11/../../../../../../x86_64-apple-darwin21/include"
ignoring nonexistent directory "/Library/Developer/CommandLineTools/SDKs/MacOSX12.sdk/usr/include"
ignoring nonexistent directory "/Library/Developer/CommandLineTools/SDKs/MacOSX12.sdk/System/Library/Frameworks"
ignoring nonexistent directory "/Library/Developer/CommandLineTools/SDKs/MacOSX12.sdk/Library/Frameworks"
#include "..." search starts here:
#include <...> search starts here:
 /Users/dstebila/Dev/OQS/liboqs/build/include
 /usr/local/Cellar/gcc/11.2.0_3/lib/gcc/11/gcc/x86_64-apple-darwin21/11/include
 /usr/local/Cellar/gcc/11.2.0_3/lib/gcc/11/gcc/x86_64-apple-darwin21/11/include-fixed
End of search list.
COLLECT_GCC_OPTIONS='-I' '/Users/dstebila/Dev/OQS/liboqs/build/include' '-fPIC' '-v' '-v' '-v' '-march=native' '-Werror' '-Wall' '-Wextra' '-Wpedantic' '-Wstrict-prototypes' '-Wshadow' '-Wformat=2' '-Wfloat-equal' '-Wwrite-strings' '-O3' '-fomit-frame-pointer' '-fdata-sections' '-ffunction-sections' '-Wbad-function-cast' '-MD' '-MT' 'src/common/sha3/xkcp_low/CMakeFiles/xkcp_low_keccakp_1600_avx2.dir/KeccakP-1600/avx2/KeccakP-1600-AVX2.S.o' '-MF' 'src/common/sha3/xkcp_low/CMakeFiles/xkcp_low_keccakp_1600_avx2.dir/KeccakP-1600/avx2/KeccakP-1600-AVX2.S.o.d' '-o' 'src/common/sha3/xkcp_low/CMakeFiles/xkcp_low_keccakp_1600_avx2.dir/KeccakP-1600/avx2/KeccakP-1600-AVX2.S.o' '-c' '-mmacosx-version-min=12.0.0'  '-dumpdir' 'src/common/sha3/xkcp_low/CMakeFiles/xkcp_low_keccakp_1600_avx2.dir/KeccakP-1600/avx2/'
 as -arch x86_64 -v -v -v -I /Users/dstebila/Dev/OQS/liboqs/build/include -force_cpusubtype_ALL -mmacosx-version-min=12.0 -defsym macOS=1 -o src/common/sha3/xkcp_low/CMakeFiles/xkcp_low_keccakp_1600_avx2.dir/KeccakP-1600/avx2/KeccakP-1600-AVX2.S.o /var/folders/xc/jknbkyc944z3m9knl6xljf0m0000gn/T//ccrgHf0V.s
Apple clang version 13.0.0 (clang-1300.0.29.30)
Target: x86_64-apple-darwin21.1.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
clang: error: no such file or directory: 'macOS=1'
```

It looks like, regardless of what compiler is used, the macOS assembler `as` is used, and `as` is really clang/llvm underneath.  

I just noticed that the [most recent commit on the XKCP repository](https://github.com/XKCP/XKCP/commit/5b81ec547d8c1453ff502e309379a99b1c43f338) is a patch to the exact file in question specifically around adjusting syntax for macOS.  Which makes me hope that perhaps if we update the XKCP code, we can remove `-Wa,-defsym,macOS=1` from our CMakeLists.txt file and then this problem will go away.

**Comment by baentsch** [comment]
> It looks like, regardless of what compiler is used, the macOS assembler as is used, and as is really clang/llvm underneath.

Might it be an option to set `CMAKE_ASM_COMPILER` explicitly then?

**Comment by dstebila** [comment]
I don't know of another assembler that even could be used.  As far as I know the command for the GNU assembler is `gas`, but that command isn't present on my Mac.

**Comment by vt-alt** [comment]
Btw, GNU assembler also could be named `as` and it's usually in `binutils` package.
```
linux$ as --version
GNU assembler (GNU Binutils) 2.37.20210908
Copyright (C) 2021 Free Software Foundation, Inc.
This program is free software; you may redistribute it under the terms of
the GNU General Public License version 3 or later.
This program has absolutely no warranty.
This assembler was configured for a target of `x86_64-alt-linux'.
```
If it's not present sometimes `gcc` itself could be used as assembler `AS=gcc`.

**Comment by baentsch** [comment]
Fixed by #1173. Thanks @jschanck !
